### PR TITLE
Have Burn calculate directly in getDamage

### PR DESCRIPTION
**Pull request by ascriptmaster**

### CODE REVIEW

**Comment by DanUgore** [comment]
Not what I meant by directly in the damage formula. There is a check specifically for burn. It is not a modifier and halves the damage.

**Comment by DanUgore** [comment]
I believe I left a comment in battle-engine.js exactly where it should be

**Comment by ascriptmaster** [comment]
I don't think I saw anything in `battle-engine.js` or `data/scripts.js` relating to it...

If you want I can push back and just leave the tests

File: test/simulator/misc/statuses.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+		assert.strictEqual(battle.p1.active[0].getStat('spe'), battle.modify(speed, 0.25));
+	});
+});
+
+describe('Toxic Poison', function () {
+	afterEach(function () {
+		battle.destroy();
+	});
+
+	it('should inflict (active turns with Toxic)/16 of max HP at the end of the turn, rounded down', function () {
```
Comment (Marty-D) [inline test/simulator/misc/statuses.js]: This should actually be something like "1/16 max HP, rounded down, multiplied by (active turns with Toxic)".

The only reason this test works is that default Blissey has 672 max HP, which just happens to be divisible by 16.

Comment (ascriptmaster) [inline test/simulator/misc/statuses.js]: RIP

I switched it to Chansey, which has 622 max HP and isn't divisible by 16.

**Comment by ascriptmaster** [comment]
Maybe I should add oldgen tests as well but I don't have a very good understanding of how they work

**Comment by DanUgore** [comment]
Burn should happen right before line 3181

**Comment by ascriptmaster** [comment]
Hmmm. Just to verify, if you do something like Level 100 Chilan Berry Steelix against a Level 1 Onix using Tackle or something, is it supposed to do 1 damage or 0 damage?

**Comment by DanUgore** [comment]
Nothing does 0 damage in Gen 6 anymore :(

**Comment by DanUgore** [comment]
In Gen 5 it'd be 0 damage

**Comment by ascriptmaster** [comment]
> In Gen 5 it'd be 0 damage
> Edit: And probably 4 and 3

With Level 1 Onix though? Damage calcs say it does flat 1 before the Chilan Berry.

**Comment by DanUgore** [comment]
Chilan cuts it to 0.

1 modified by 0x800 (0.5) = 0.5, round 0.5 down to 0

**Comment by ascriptmaster** [comment]
So I have to add ANOTHER check specific to Gen 6 that after `ModifyDamage` is run to reset things back to 1 damage if needed as well?

**Comment by DanUgore** [comment]
after final modifiers yea. just check if damage is 0 and if it is make it 1... I THINK

**Comment by ascriptmaster** [comment]
Let's just wait for @Marty-D to dispense his infinite wisdom then.

**Comment by DanUgore** [comment]
Bruh. It's a simple check and hell it might even be there already.



-- 
-- Hugh Gordon

**Comment by Marty-D** [comment]
> Hmmm. Just to verify, if you do something like Level 100 Chilan Berry Steelix against a Level 1 Onix using Tackle or something, is it supposed to do 1 damage or 0 damage?

It's 0 damage in Gen 5 only. Every other gen ensures minimum damage is 1 after processing other stuff. Even Gen 5 does it, but for whatever reason the final modifiers (screens, damage-modifying Berries, Multiscale, etc.) are done after that step. It was fixed in Gen 6 to come before.

**Comment by ascriptmaster** [comment]
How is the Gen 4 and under damage calc completely different, pray tell?

**Comment by ascriptmaster** [comment]
While I wait for Marty's reply, here's a conceptual of how burn would work within the damage function. I'll move the thing about forcing damage to 1 later.

**Comment by Joimer** [comment]
Gens 1 and 2  actually do this with burn, but in an entirely different way.

**Comment by ascriptmaster** [comment]
The issue I'm worried about at this point is @Marty-D's comment:

> (Prior to Gen 5 the damage formula was completely different anyway, but that's another problem.)

If I'm going to make a pull request to fix damage calculations, I'm going to do it right :\

**Comment by Marty-D** [comment]
@ascriptmaster: [Gen 4](http://www.smogon.com/dp/articles/damage_formula) compared to [Gen 5](http://www.smogon.com/bw/articles/bw_complete_damage_formula), for example. Probably beyond the scope of this pull request.

**Comment by ascriptmaster** [comment]
All right then. @Marty-D and @DanUgore, is this the proper way to do burn?

**Comment by Zarel** [comment]
> is this the proper way to do burn?

> build error

**Comment by ascriptmaster** [comment]
> `1) Battle simulation Moves Thousand Arrows should hit non-Flying-type Pokemon with Levitate with standard type effectiveness`

Hahaha wow. This shouldn't be failing at all, even occasionally, I'll have to look into why it'd fail in the first place.

**Comment by ascriptmaster** [comment]
Build error from the Mocha test is to be fixed in #1806

**Comment by ascriptmaster** [comment]
Test was fixed, so this one now passes.

**Comment by Joimer** [comment]
It's pretty amusing that an error that happened very few times happened on the Travis Build.
Anyway, seems fine. If you want to add tests for past gens, the way to go is similar, only the mechanic is not the same, when calculating the stat, attack is halved.

In fact, gen 1 is more complex than that, since the halving can be re-applied, but that's an entirely different thing that is needing a refactor that I'm working on.

**Comment by ascriptmaster** [comment]
The error happened very few times on the Travis build because it was an edge case when the RNG rolled a crit of Thousand Arrows (6.25% chance). It was notoriously hard for me to reproduce the failure in the first place in order to fix it.

**Pull request merged**